core/src/main/java/com/linecorp/armeria/server/conditional/ConditionalRequestUtil.java

```diff
@@ -0,0 +1,202 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.conditional;
+
+import static com.linecorp.armeria.server.conditional.ETag.parseHeader;
+import static com.linecorp.armeria.server.conditional.ETagResponse.PERFORM_METHOD;
+import static com.linecorp.armeria.server.conditional.ETagResponse.SKIP_METHOD_NOT_MODIFIED;
+import static com.linecorp.armeria.server.conditional.ETagResponse.SKIP_METHOD_PRECONDITION_FAILED;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.math.LongMath;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * This class helps with handling of Conditional Requests as per
+ * <a href="https://datatracker.ietf.org/doc/html/rfc2732">RFC 7232</a>.
+ */
+public final class ConditionalRequestUtil {
+    private ConditionalRequestUtil() {}
+
+    @VisibleForTesting
+    static boolean strongComparison(List<ETag> eTags, ETag dataETag) {
+        if (dataETag.isWeak()) {
+            return false;
+        }
+        for (ETag eTag : eTags) {
+            if (eTag.isStrong() && eTag.getETag().equals(dataETag.getETag())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @VisibleForTesting
+    static boolean weakComparison(List<ETag> requestETags, ETag dataETag) {
+        for (ETag eTag : requestETags) {
+            if (eTag.getETag().equals(dataETag.getETag())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Performs an If-Match.
+     * @param requestETags The parsed If-None-Match header.
+     * @param dataETag The ETag of the current data. If the resource doesn't exists, this is null.
+     *      This is mostly used to make sure PUT operations don't overwrite existing data.
+     * @return true if the method should be performed (ie. NOT just return 304 Not Modified).
+     */
+    public static ETagResponse ifMatch(List<ETag> requestETags, @Nullable ETag dataETag) {
+        requireNonNull(requestETags, "requestETags cannot be null");
+        // Making sure that it's the same instance
+        if (requestETags == ETag.ASTERISK_ETAG) {
+            if (dataETag != null) {
+                return PERFORM_METHOD;
+            } else {
+                return SKIP_METHOD_PRECONDITION_FAILED;
+            }
+        }
+        if (dataETag == null) {
+            return PERFORM_METHOD;
+        }
+        // If-Match requires strong match which can never match weak tags
+        if (strongComparison(requestETags, dataETag)) {
+            return SKIP_METHOD_PRECONDITION_FAILED;
+        }
+        return PERFORM_METHOD;
```

Comment:
Condition is flipped

```suggestion
        if (strongComparison(requestETags, dataETag)) {
            return PERFORM_METHOD;
        }
        return SKIP_METHOD_PRECONDITION_FAILED;
```